### ISSUE TITLE
patch: perl 5.8.0 compile error on ia64/HP-UX (broken attribute test)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6506]
Migrated from [rt.perl.org#22189](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22189) (status was 'resolved')

Searchable as RT22189$

**Comment by p5pRT** [comment]
### From pere@hungry.com


I ran into a compile error when trying to compile perl 5\.8\.0 on ia64
HP/UX 11\.22\.  I got this error message when trying to compile
miniperlmain\.c&#8203;:

&nbsp;  Error 172&#8203;: "perlio\.h"\, line 209 \# Undeclared variable '\_\_printf\_\_'\.

The problem was the following code&#8203;:

&nbsp;    \_\_attribute\_\_ \(\(\_\_format\_\_\(\_\_printf\_\_\, 1\, 2\)\)\);

Tracking the problem further\, I discovered the test for \_\_attribute\_\_
in Configure\, where it is testing if the following construct is
compilable&#8203;:

&nbsp;  \#include \<stdio\.h>
&nbsp;  void croak \(char\* pat\,\.\.\.\) \_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);

I also discovered that the result from this test isn't printed to
stdout\, making it hard to find out what the result from this test was\.

I belive the test in Configure should test for the same feature that
is used in the perl code\, and rewrote 'format' to '\_\_format\_\_' and
'printf' to '\_\_printf\_\_' to match perlio\.h\.  I also printed the result
of the test to stdout\, making it easy see that this test now fails as
it should on HP/UX\.

With the following patch\, I got past the previous problem area\.
Please include this patch in the next version of perl\.

<details><summary>Inline Patch</summary>

```diff
diff -ur src-5.8.0/Configure src-5.8.0-local/Configure
--- src-5.8.0/Configure	2002-07-19 00:55:15.000000000 +0200
+++ src-5.8.0-local/Configure	2003-05-13 20:23:30.000000000 +0200
@@ -9382,18 +9382,18 @@
 echo "Checking whether your compiler can handle __attribute__ ..." >&4
 $cat >attrib.c <<'EOCP'
 #include <stdio.h>
-void croak (char* pat,...) __attribute__((format(printf,1,2),noreturn));
+void croak (char* pat,...) __attribute__((__format__(__printf__,1,2),noreturn));
 EOCP
 if $cc $ccflags -c attrib.c >attrib.out 2>&1 ; then
 	if $contains 'warning' attrib.out >/dev/null 2>&1; then
-		echo "Your C compiler doesn't fully support __attribute__."
+		echo "Your C compiler doesn't fully support __attribute__." >&4
 		val="$undef"
 	else
-		echo "Your C compiler supports __attribute__."
+		echo "Your C compiler supports __attribute__." >&4
 		val="$define"
 	fi
 else
-	echo "Your C compiler doesn't seem to understand __attribute__ at all."
+	echo "Your C compiler doesn't seem to understand __attribute__ at all." >&4
 	val="$undef"
 fi
 set d_attribut
```
</details>

**Comment by p5pRT** [comment]
### From @doughera88

On Tue\, 13 May 2003\, Petter Reinholdtsen wrote&#8203;:

> I ran into a compile error when trying to compile perl 5\.8\.0 on ia64
> HP/UX 11\.22\.  I got this error message when trying to compile
> miniperlmain\.c&#8203;:
>
>   Error 172&#8203;: "perlio\.h"\, line 209 \# Undeclared variable '\_\_printf\_\_'\.
>
> The problem was the following code&#8203;:
>
>     \_\_attribute\_\_ \(\(\_\_format\_\_\(\_\_printf\_\_\, 1\, 2\)\)\);
>
> Tracking the problem further\, I discovered the test for \_\_attribute\_\_
> in Configure\, where it is testing if the following construct is
> compilable&#8203;:
>
>   \#include \<stdio\.h>
>   void croak \(char\* pat\,\.\.\.\) \_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);

> I belive the test in Configure should test for the same feature that
> is used in the perl code\, and rewrote 'format' to '\_\_format\_\_' and
> 'printf' to '\_\_printf\_\_' to match perlio\.h\.

Unfortunately\, both forms \(format and \_\_format\_\_\) are used in the perl
code\.  I don't know why\.  Nor do I understand why your compiler would
accept one but not the other\.  That's truly strange\, and probably worth
investigating\.  Does your compiler actually support
\_\_attribute\_\_\(\(format\(\)\)\) but not \_\_attribute\_\_\(\(\_\_format\_\_\(\)\)\)\, or is the
test flawed in some other way such that \_\_attribute\_\_ isn't supported\, but
we fail to adequately detect that\.

> I also discovered that the result from this test isn't printed to
> stdout\, making it hard to find out what the result from this test was\.

Actually\, it does print the result\.  You just told Configure to suppress
that information by supplying the \-s switch to Configure\.  Configure
similarly suppresses other outputs from the "Checking \.\.\." sorts of tests\.
If you don't want Configure to suppress information\, don't supply the \-s
switch\.

In short\, I have no objection to changing format to \_\_format\_\_ and
printf to \_\_printf\_\_\, but it'd be nice to know just why it matters so that
we can make the test more robust so that this issue wouldn't have
come up in the first place\.

\-\- 
&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu

**Comment by p5pRT** [comment]
### From Robin.Barker@npl.co.uk

Hi

I changed 'printf' to '\_\_printf\_\_' in the format enabled by CHECKFORMAT
because the 'printf' was caught up in stdio/sfio MACRO merry\-go\-round\.

I could have changed format but it wasn't a problem\, and I didn't know 
it was an option\.

I think the remaining 'format's and 'printf's should be changed to
'\_\_format\_\_' and '\_\_printf\_\_'\, including in Configure\.

Robin

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Andy Dougherty \[mailto&#8203;:doughera@&#8203;lafayette\.edu\]
Sent&#8203;: 14 May 2003 15&#8203;:26
To&#8203;: perl5\-porters@&#8203;perl\.org
Subject&#8203;: Re&#8203;: \[perl \#22189\] patch&#8203;: perl 5\.8\.0 compile error on ia64/HP\-UX
\(broken attribute test\)


On Tue\, 13 May 2003\, Petter Reinholdtsen wrote&#8203;:

> I ran into a compile error when trying to compile perl 5\.8\.0 on ia64
> HP/UX 11\.22\.  I got this error message when trying to compile
> miniperlmain\.c&#8203;:
>
>   Error 172&#8203;: "perlio\.h"\, line 209 \# Undeclared variable '\_\_printf\_\_'\.
>
> The problem was the following code&#8203;:
>
>     \_\_attribute\_\_ \(\(\_\_format\_\_\(\_\_printf\_\_\, 1\, 2\)\)\);
>
> Tracking the problem further\, I discovered the test for \_\_attribute\_\_
> in Configure\, where it is testing if the following construct is
> compilable&#8203;:
>
>   \#include \<stdio\.h>
>   void croak \(char\* pat\,\.\.\.\)
\_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);

> I belive the test in Configure should test for the same feature that
> is used in the perl code\, and rewrote 'format' to '\_\_format\_\_' and
> 'printf' to '\_\_printf\_\_' to match perlio\.h\.

Unfortunately\, both forms \(format and \_\_format\_\_\) are used in the perl
code\.  I don't know why\.  Nor do I understand why your compiler would
accept one but not the other\.  That's truly strange\, and probably worth
investigating\.  Does your compiler actually support
\_\_attribute\_\_\(\(format\(\)\)\) but not \_\_attribute\_\_\(\(\_\_format\_\_\(\)\)\)\, or is
the
test flawed in some other way such that \_\_attribute\_\_ isn't supported\,
but
we fail to adequately detect that\.

> I also discovered that the result from this test isn't printed to
> stdout\, making it hard to find out what the result from this test was\.

Actually\, it does print the result\.  You just told Configure to suppress
that information by supplying the \-s switch to Configure\.  Configure
similarly suppresses other outputs from the "Checking \.\.\." sorts of
tests\.
If you don't want Configure to suppress information\, don't supply the \-s
switch\.

In short\, I have no objection to changing format to \_\_format\_\_ and
printf to \_\_printf\_\_\, but it'd be nice to know just why it matters so
that
we can make the test more robust so that this issue wouldn't have
come up in the first place\.

\-\- 
&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu



---
This e\-mail and any attachments may contain confidential and/or
privileged material; it is for the intended addressee\(s\) only\.
If you are not a named addressee\, you must not use\, retain or
disclose such information\.

NPL Management Ltd cannot guarantee that the e\-mail or any
attachments are free from viruses\.

NPL Management Ltd\. Registered in England and Wales\. No&#8203;: 2937881
Registered Office&#8203;: Teddington\, Middlesex\, United Kingdom TW11 0LW\.


---

**Comment by p5pRT** [comment]
### From pere@hungry.com


\[Andy Dougherty\]
> Unfortunately\, both forms \(format and \_\_format\_\_\) are used in the perl
> code\.  I don't know why\.  Nor do I understand why your compiler would
> accept one but not the other\.  That's truly strange\, and probably worth
> investigating\.  Does your compiler actually support
> \_\_attribute\_\_\(\(format\(\)\)\) but not \_\_attribute\_\_\(\(\_\_format\_\_\(\)\)\)\, or is the
> test flawed in some other way such that \_\_attribute\_\_ isn't supported\, but
> we fail to adequately detect that\.

It is the other way around\.  It seem to support printf\, but not
\_\_printf\_\_\.  Or perhaps it uses the function pointer to printf\(\)?

Given the two files  foo\.c and foo2\.c\, I get this behavoiur&#8203;:

&nbsp;  % cat foo\.c
&nbsp;  \#include \<stdio\.h>
&nbsp;  void croak \(char\* pat\,\.\.\.\) \_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);
&nbsp;  % cat foo2\.c
&nbsp;  \#include \<stdio\.h>
&nbsp;  void croak \(char\* pat\,\.\.\.\) \_\_attribute\_\_\(\(\_\_format\_\_\(\_\_printf\_\_\,1\,2\)\,noreturn\)\);
&nbsp;  % cc\-wrapper \-c foo\.c
&nbsp;  % cc\-wrapper \-c foo2\.c
&nbsp;  Error 172&#8203;: "foo2\.c"\, line 2 \# Undeclared variable '\_\_printf\_\_'\.
&nbsp;    void croak \(char\* pat\,\.\.\.\) \_\_attribute\_\_\(\(\_\_format\_\_\(\_\_printf\_\_\,1\,2\)\,noretu
&nbsp;                                                         ^^^^^^^^^^
&nbsp;  %

Thus the Configure test fail\.

&nbsp;  % uname \-a; cc \-V
&nbsp;  HP\-UX unknown B\.11\.22 U ia64 unknown unknown HP\-UX
&nbsp;  cc&#8203;: HP aC\+\+/ANSI C B3910B A\.05\.41 \[Nov 1 2002\]

> In short\, I have no objection to changing format to \_\_format\_\_ and
> printf to \_\_printf\_\_\, but it'd be nice to know just why it matters
> so that we can make the test more robust so that this issue wouldn't
> have come up in the first place\.

I do not know why it matters\, but it does\.  I suspect it is safer to
use \_\_format\_\_ and \_\_printf\_\_ in the source when this feature is used\,
to make sure the correct feature is used\.

> Actually\, it does print the result\.  You just told Configure to
> suppress that information by supplying the \-s switch to Configure\.

Right\.  Good point\.  I didn't write the compile scripts I am using\,
and did not notice that switch\.  It is now removed\, and that part of
the patch is removed from my version\.  Thanks for the tip\.

**Comment by p5pRT** [comment]
### From Robin.Barker@npl.co.uk

The format attributes are not normally visible to the compiler\.
Unless you are using \-DCHECK\_FORMAT\, the \_\_printf\_\_ are not seen\.
The attribute setting in perlio\.h were not protected by 
C\<ifdef CHECK\_FORMAT> \-\- this is a bug and was fixed by my
patch 18556\.

Robin

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Petter Reinholdtsen \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\]
Sent&#8203;: 13 May 2003 19&#8203;:42
To&#8203;: perl5\-porters@&#8203;perl\.org
Subject&#8203;: \[perl \#22189\] patch&#8203;: perl 5\.8\.0 compile error on ia64/HP\-UX
\(broken attribute test\)


\# New Ticket Created by  Petter Reinholdtsen 
\# Please include the string&#8203;:  \[perl \#22189\]
\# in the subject line of all future correspondence about this issue\. 
\# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=22189 >



I ran into a compile error when trying to compile perl 5\.8\.0 on ia64
HP/UX 11\.22\.  I got this error message when trying to compile
miniperlmain\.c&#8203;:

&nbsp;  Error 172&#8203;: "perlio\.h"\, line 209 \# Undeclared variable '\_\_printf\_\_'\.

The problem was the following code&#8203;:

&nbsp;    \_\_attribute\_\_ \(\(\_\_format\_\_\(\_\_printf\_\_\, 1\, 2\)\)\);

Tracking the problem further\, I discovered the test for \_\_attribute\_\_
in Configure\, where it is testing if the following construct is
compilable&#8203;:

&nbsp;  \#include \<stdio\.h>
&nbsp;  void croak \(char\* pat\,\.\.\.\)
\_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);

I also discovered that the result from this test isn't printed to
stdout\, making it hard to find out what the result from this test was\.

I belive the test in Configure should test for the same feature that
is used in the perl code\, and rewrote 'format' to '\_\_format\_\_' and
'printf' to '\_\_printf\_\_' to match perlio\.h\.  I also printed the result
of the test to stdout\, making it easy see that this test now fails as
it should on HP/UX\.

With the following patch\, I got past the previous problem area\.
Please include this patch in the next version of perl\.

<details><summary>Inline Patch</summary>

```diff
diff -ur src-5.8.0/Configure src-5.8.0-local/Configure
--- src-5.8.0/Configure	2002-07-19 00:55:15.000000000 +0200
+++ src-5.8.0-local/Configure	2003-05-13 20:23:30.000000000 +0200
@@ -9382,18 +9382,18 @@
 echo "Checking whether your compiler can handle __attribute__ ..." >&4
 $cat >attrib.c <<'EOCP'
 #include <stdio.h>
-void croak (char* pat,...)

```
</details>
__attribute__((format(printf,1,2),noreturn));
\+void croak \(char\* pat\,\.\.\.\)
\_\_attribute\_\_\(\(\_\_format\_\_\(\_\_printf\_\_\,1\,2\)\,noreturn\)\);
 EOCP
 if $cc $ccflags \-c attrib\.c >attrib\.out 2>&1 ; then
&nbsp;     if $contains 'warning' attrib\.out >/dev/null 2>&1; then
\-        echo "Your C compiler doesn't fully support
\_\_attribute\_\_\."
\+        echo "Your C compiler doesn't fully support
\_\_attribute\_\_\." >&4
&nbsp;         val="$undef"
&nbsp;     else
\-        echo "Your C compiler supports \_\_attribute\_\_\."
\+        echo "Your C compiler supports \_\_attribute\_\_\." >&4
&nbsp;         val="$define"
&nbsp;     fi
 else
\-    echo "Your C compiler doesn't seem to understand \_\_attribute\_\_
at all\."
\+    echo "Your C compiler doesn't seem to understand \_\_attribute\_\_
at all\." >&4
&nbsp;     val="$undef"
 fi
 set d\_attribut



---
This e\-mail and any attachments may contain confidential and/or
privileged material; it is for the intended addressee\(s\) only\.
If you are not a named addressee\, you must not use\, retain or
disclose such information\.

NPL Management Ltd cannot guarantee that the e\-mail or any
attachments are free from viruses\.

NPL Management Ltd\. Registered in England and Wales\. No&#8203;: 2937881
Registered Office&#8203;: Teddington\, Middlesex\, United Kingdom TW11 0LW\.


---

**Comment by p5pRT** [comment]
### From @jhi

I applied other parts of the patch except for the MIME&#8203;::Base64 parts that I forwarded to 
Gisle\, he will decide what to do with those\.  Therefore I'm marking this problem ticket as 
resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Robin Barker \<Robin\.Barker@&#8203;npl\.co\.uk> writes&#8203;:
>Hi
>
>I changed 'printf' to '\_\_printf\_\_' in the format enabled by CHECKFORMAT
>because the 'printf' was caught up in stdio/sfio MACRO merry\-go\-round\.
>
>I could have changed format but it wasn't a problem\, and I didn't know 
>it was an option\.
>
>I think the remaining 'format's and 'printf's should be changed to
>'\_\_format\_\_' and '\_\_printf\_\_'\, including in Configure\.
>
>Robin

That depends if HP/UX really undestands the forms without the \_\_
then why should we limit ourselves to the gcc\-ish \_\_ forms?



>
>\-\-\-\-\-Original Message\-\-\-\-\-
>From&#8203;: Andy Dougherty \[mailto&#8203;:doughera@&#8203;lafayette\.edu\]
>Sent&#8203;: 14 May 2003 15&#8203;:26
>To&#8203;: perl5\-porters@&#8203;perl\.org
>Subject&#8203;: Re&#8203;: \[perl \#22189\] patch&#8203;: perl 5\.8\.0 compile error on ia64/HP\-UX
>\(broken attribute test\)
>
>
>On Tue\, 13 May 2003\, Petter Reinholdtsen wrote&#8203;:
>
>> I ran into a compile error when trying to compile perl 5\.8\.0 on ia64
>> HP/UX 11\.22\.  I got this error message when trying to compile
>> miniperlmain\.c&#8203;:
>>
>>   Error 172&#8203;: "perlio\.h"\, line 209 \# Undeclared variable '\_\_printf\_\_'\.
>>
>> The problem was the following code&#8203;:
>>
>>     \_\_attribute\_\_ \(\(\_\_format\_\_\(\_\_printf\_\_\, 1\, 2\)\)\);
>>
>> Tracking the problem further\, I discovered the test for \_\_attribute\_\_
>> in Configure\, where it is testing if the following construct is
>> compilable&#8203;:
>>
>>   \#include \<stdio\.h>
>>   void croak \(char\* pat\,\.\.\.\)
>\_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);
>
>> I belive the test in Configure should test for the same feature that
>> is used in the perl code\, and rewrote 'format' to '\_\_format\_\_' and
>> 'printf' to '\_\_printf\_\_' to match perlio\.h\.
>
>Unfortunately\, both forms \(format and \_\_format\_\_\) are used in the perl
>code\.  I don't know why\.  Nor do I understand why your compiler would
>accept one but not the other\.  That's truly strange\, and probably worth
>investigating\.  Does your compiler actually support
>\_\_attribute\_\_\(\(format\(\)\)\) but not \_\_attribute\_\_\(\(\_\_format\_\_\(\)\)\)\, or is
>the
>test flawed in some other way such that \_\_attribute\_\_ isn't supported\,
>but
>we fail to adequately detect that\.
>
>> I also discovered that the result from this test isn't printed to
>> stdout\, making it hard to find out what the result from this test was\.
>
>Actually\, it does print the result\.  You just told Configure to suppress
>that information by supplying the \-s switch to Configure\.  Configure
>similarly suppresses other outputs from the "Checking \.\.\." sorts of
>tests\.
>If you don't want Configure to suppress information\, don't supply the \-s
>switch\.
>
>In short\, I have no objection to changing format to \_\_format\_\_ and
>printf to \_\_printf\_\_\, but it'd be nice to know just why it matters so
>that
>we can make the test more robust so that this issue wouldn't have
>come up in the first place\.
>
>\-\- 
>    Andy Dougherty        doughera@&#8203;lafayette\.edu
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>This e\-mail and any attachments may contain confidential and/or
>privileged material; it is for the intended addressee\(s\) only\.
>If you are not a named addressee\, you must not use\, retain or
>disclose such information\.
>
>NPL Management Ltd cannot guarantee that the e\-mail or any
>attachments are free from viruses\.
>
>NPL Management Ltd\. Registered in England and Wales\. No&#8203;: 2937881
>Registered Office&#8203;: Teddington\, Middlesex\, United Kingdom TW11 0LW\.
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @tux

On Sat 17 May 2003 19&#8203;:39\, Nick Ing\-Simmons \<nick@&#8203;ing\-simmons\.net> wrote&#8203;:
> Robin Barker \<Robin\.Barker@&#8203;npl\.co\.uk> writes&#8203;:
> >Hi
> >
> >I changed 'printf' to '\_\_printf\_\_' in the format enabled by CHECKFORMAT
> >because the 'printf' was caught up in stdio/sfio MACRO merry\-go\-round\.
> >
> >I could have changed format but it wasn't a problem\, and I didn't know 
> >it was an option\.
> >
> >I think the remaining 'format's and 'printf's should be changed to
> >'\_\_format\_\_' and '\_\_printf\_\_'\, including in Configure\.
> >
> >Robin
> 
> That depends if HP/UX really undestands the forms without the \_\_
> then why should we limit ourselves to the gcc\-ish \_\_ forms?

What should I test? \(HP\-UX 10\.20 and 11\.00 available\)

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.0 & 633 on HP\-UX 10\.20 & 11\.00\, AIX 4\.2\, AIX 4\.3\,
&nbsp;  WinNT 4\, Win2K pro & WinCE 2\.11\.  Smoking perl CORE&#8203;: smokers@&#8203;perl\.org
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From Robin.Barker@npl.co.uk

Nick

Something needs to change \-\- at the moment
nostdio\.h defines 'printf' as 'CANNOT \_printf\_'\, so
'format\(printf\,1\,2\)' becomes 'format\(CANNOT \_printf\_\,1\,2\)'
which does not parse\.  \(You only get this with \-DCHECK\_FORMAT\.\)

Using '\_\_printf\_\_' seemed the neated way to fix this\.

Robin

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Nick Ing\-Simmons \[mailto&#8203;:nick@&#8203;ing\-simmons\.net\]
Sent&#8203;: 17 May 2003 18&#8203;:39
To&#8203;: Robin\.Barker@&#8203;npl\.co\.uk
Cc&#8203;: 'Andy Dougherty'; perl5\-porters@&#8203;perl\.org
Subject&#8203;: RE&#8203;: \[perl \#22189\] patch&#8203;: perl 5\.8\.0 compile error on ia64/HP\-UX
\( broken attribute test\)


Robin Barker \<Robin\.Barker@&#8203;npl\.co\.uk> writes&#8203;:
>Hi
>
>I changed 'printf' to '\_\_printf\_\_' in the format enabled by CHECKFORMAT
>because the 'printf' was caught up in stdio/sfio MACRO merry\-go\-round\.
>
>I could have changed format but it wasn't a problem\, and I didn't know 
>it was an option\.
>
>I think the remaining 'format's and 'printf's should be changed to
>'\_\_format\_\_' and '\_\_printf\_\_'\, including in Configure\.
>
>Robin

That depends if HP/UX really undestands the forms without the \_\_
then why should we limit ourselves to the gcc\-ish \_\_ forms?



>
>\-\-\-\-\-Original Message\-\-\-\-\-
>From&#8203;: Andy Dougherty \[mailto&#8203;:doughera@&#8203;lafayette\.edu\]
>Sent&#8203;: 14 May 2003 15&#8203;:26
>To&#8203;: perl5\-porters@&#8203;perl\.org
>Subject&#8203;: Re&#8203;: \[perl \#22189\] patch&#8203;: perl 5\.8\.0 compile error on
ia64/HP\-UX
>\(broken attribute test\)
>
>
>On Tue\, 13 May 2003\, Petter Reinholdtsen wrote&#8203;:
>
>> I ran into a compile error when trying to compile perl 5\.8\.0 on ia64
>> HP/UX 11\.22\.  I got this error message when trying to compile
>> miniperlmain\.c&#8203;:
>>
>>   Error 172&#8203;: "perlio\.h"\, line 209 \# Undeclared variable '\_\_printf\_\_'\.
>>
>> The problem was the following code&#8203;:
>>
>>     \_\_attribute\_\_ \(\(\_\_format\_\_\(\_\_printf\_\_\, 1\, 2\)\)\);
>>
>> Tracking the problem further\, I discovered the test for \_\_attribute\_\_
>> in Configure\, where it is testing if the following construct is
>> compilable&#8203;:
>>
>>   \#include \<stdio\.h>
>>   void croak \(char\* pat\,\.\.\.\)
>\_\_attribute\_\_\(\(format\(printf\,1\,2\)\,noreturn\)\);
>
>> I belive the test in Configure should test for the same feature that
>> is used in the perl code\, and rewrote 'format' to '\_\_format\_\_' and
>> 'printf' to '\_\_printf\_\_' to match perlio\.h\.
>
>Unfortunately\, both forms \(format and \_\_format\_\_\) are used in the perl
>code\.  I don't know why\.  Nor do I understand why your compiler would
>accept one but not the other\.  That's truly strange\, and probably worth
>investigating\.  Does your compiler actually support
>\_\_attribute\_\_\(\(format\(\)\)\) but not \_\_attribute\_\_\(\(\_\_format\_\_\(\)\)\)\, or is
>the
>test flawed in some other way such that \_\_attribute\_\_ isn't supported\,
>but
>we fail to adequately detect that\.
>
>> I also discovered that the result from this test isn't printed to
>> stdout\, making it hard to find out what the result from this test
was\.
>
>Actually\, it does print the result\.  You just told Configure to
suppress
>that information by supplying the \-s switch to Configure\.  Configure
>similarly suppresses other outputs from the "Checking \.\.\." sorts of
>tests\.
>If you don't want Configure to suppress information\, don't supply the
\-s
>switch\.
>
>In short\, I have no objection to changing format to \_\_format\_\_ and
>printf to \_\_printf\_\_\, but it'd be nice to know just why it matters so
>that
>we can make the test more robust so that this issue wouldn't have
>come up in the first place\.
>
>\-\- 
>    Andy Dougherty        doughera@&#8203;lafayette\.edu
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>This e\-mail and any attachments may contain confidential and/or
>privileged material; it is for the intended addressee\(s\) only\.
>If you are not a named addressee\, you must not use\, retain or
>disclose such information\.
>
>NPL Management Ltd cannot guarantee that the e\-mail or any
>attachments are free from viruses\.
>
>NPL Management Ltd\. Registered in England and Wales\. No&#8203;: 2937881
>Registered Office&#8203;: Teddington\, Middlesex\, United Kingdom TW11 0LW\.
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/



---
This e\-mail and any attachments may contain confidential and/or
privileged material; it is for the intended addressee\(s\) only\.
If you are not a named addressee\, you must not use\, retain or
disclose such information\.

NPL Management Ltd cannot guarantee that the e\-mail or any
attachments are free from viruses\.

NPL Management Ltd\. Registered in England and Wales\. No&#8203;: 2937881
Registered Office&#8203;: Teddington\, Middlesex\, United Kingdom TW11 0LW\.


---

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Robin Barker \<Robin\.Barker@&#8203;npl\.co\.uk> writes&#8203;:
>Nick
>
>Something needs to change 

Apparently\!

>\-\- at the moment
>nostdio\.h defines 'printf' as 'CANNOT \_printf\_'\, so
>'format\(printf\,1\,2\)' becomes 'format\(CANNOT \_printf\_\,1\,2\)'
>which does not parse\.  \(You only get this with \-DCHECK\_FORMAT\.\)

For some meaning of "at the moment"\. I am reasonably sure 
we fixed the \#define of printf in bleadperl \(post 5\.8\.0\)
It does not only mess up attributes but also preculdes using printf\(\) 
in XS code in general \- in the brave new PerlIO world that is allowed\.

In older perls the ordering of the \#include\-s via perl\.h is supposed to 
be cleverly \(diabolically?\) contrived so that the attributes worked\.
I know that when that printf mess went in the equivalent of 
\-DCHECK\_FORMAT was the default \(at least for gcc I habitually used\)\.
I guess that now CHECK\_FORMAT is optional \(because of all the %\_ stuff
makes it noisy\) \- patches may have broken the clever ordering and 
we only just noticed\.

>


>Using '\_\_printf\_\_' seemed the neated way to fix this\.

But suppose \(as seems to be the case\) HP/UX compiler supports
format\(printf\,\.\.\.\) but does NOT support format\(\_\_printf\_\_\,\.\.\.\)?
Do we want to stop HP/UXers do \-DCHECK\_FORMAT?

The easiest way is 
\#undef printf

the best way is loose the \#define\.


\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @doughera88

On Sat\, 17 May 2003\, Nick Ing\-Simmons wrote&#8203;:

> >I think the remaining 'format's and 'printf's should be changed to
> >'\_\_format\_\_' and '\_\_printf\_\_'\, including in Configure\.

I agree with that\.  And it's certainly clear that Configure ought
to be testing whatever perl ends up actually using\.

> That depends if HP/UX really undestands the forms without the \_\_
> then why should we limit ourselves to the gcc\-ish \_\_ forms?

It wasn't clear to me that HP/UX was actually doing anything useful with
the "attribute"\.  If I recall correctly\, the original poster was using a
compiler called "cc\-wrapper"\.  I have no idea exactly what that "wrapper"
is actually doing\.

More to the point\, I think we have plenty of gcc users on p5p anyway who
can catch printf formatting errors; it's probably not worth too much
effort to expand the checking further\.

\-\- 
&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu

**Comment by p5pRT** [comment]
### From pere@hungry.com

\[Andy Dougherty\]
> If I recall correctly\, the original poster was using a compiler
> called "cc\-wrapper"\.  I have no idea exactly what that "wrapper" is
> actually doing\.

It just make sure we can use the same compiler setting on all our 9
unix architectures\.  On HP/UX\, it turns into this&#8203;:

&nbsp;  PATH=/opt/ansic/bin&#8203;:$PATH cc \+DD64 \-Ae \-O \-I/local/include \-L/local/lib

**Comment by p5pRT** [comment]
### From @jhi

I believe this problem got fixed\, and am therefore marking the problem ticket
as resolved\.  Perl 5\.8\.1 Release Candidates should start coming out soon\-ish\,
stay tuned\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'
